contribute/design/1.x/replica_scaleup/dataplane/20190910-replica-scaleup.md

```diff
@@ -0,0 +1,278 @@
+cstor-volume-mgmt is the container that fills the RF and CF details into
+istgt.conf file by reading from CStorVolume CR.
+
+Another property that cstor-istgt uses is 'quorum'. This property is set at
+replica, and, replica sends this to istgt during handshake phase.
+This property can take values of 'on' and 'off'.
+'on' means that data written to this replica is available for reads.
+'off' means that data written to this replica is lost.
+cstor-pool-mgmt is the container that sets quorum property at replica.
+It sets property as 'on' if replica is created first time related to that CVR.
+Otherwise, it sets to 'off'.
+
+#### Current Implementation -- Config
+- cstor-volume-mgmt reads configuration from CStorVolume CR.
+  - spec.replicationFactor
+  - spec.consistencyFactor
+- cstor-volume-mgmt updates the above info into istgt.conf file as
+  - ReplicationFactor
+  - ConsistencyFactor
+- Based on the status of CVR and availability of replica, cstor-pool-mgmt
+sets quorum property while creating replica.
+
+#### Current Implementation
+Above mentioned properties are used as below at istgt:
+- Allow replicas (upto max of 5) to connect to target until RF number of
+healthy replicas are available.
+- Atleast CF in-quorum replicas are required to perform rebuilding and start IOs
+- Rebuilding will be triggered and replicas become healthy. As RF number of
+replicas are healthy, other connected replicas, if any, gets disconnected.
+
+#### Shortcomings With Current Implementation
+- Due to misconfiguration from user, if old replica connects back instead of
+the replaced one, there are chances of serving wrong data, and can lead to
+data corruption.
+- There is no data-consistent way of increasing RF and CF of CStorVolume CR.
+
+### Proposed Implementation
+Increase in replication factor will be provided in declarative way.
+For replicas identification that can connect to target, replicas information
+as well need to be stored in CStorVolume CR and istgt.conf.
+
+New fields will be added to CStorVolume CR and istgt.conf.
+In CStorVolume CR, spec.DesiredReplicationFactor will be added to help in
+adding replicas, and, status.ReplicaList to store list of known replicas.
+In istgt.conf, DesiredReplicationFactor and ReplicaList will be added.
+
+
+#### Adding new replica
+- User will create proper CVR with other than Empty status.phase
+- User will edit CStorVolume CR to set 'DesiredReplicationFactor'
+
+#### Replacing non-existing replica
+Assuming user has reduced RF, CF of CStorVolume CR to have volume in Online
+state, leaving the DesiredReplicationFactor as it is will increase the RF.
+
+### High Level Design
+cstor-volume-mgmt watches for CStorVolume CR and updates istgt.conf file if
+there is any change in DesiredReplicationFactor.
+It will trigger istgtcontrol command so that istgt updates DRF during runtime.
+
+A listener will be created on UnixDomain socket in cstor-volume-mgmt container.
```

Comment:
Is this new socket or will this reuse existing channel?

Reply:
This would be new one. Listener should be on cstor-volume-mgmt